### PR TITLE
Improve `mypy` type checking

### PR DESCRIPTION
**Pull request by asnare**
## Changes

This PR:

 - Adds type hints to functions/methods that Mypy was skipping; by default it will skip checking for functions/methods if the signature doesn't include in any hints.
 - Fixes some type-checking errors that turned up once Mypy started checking previously skipped code.

### Tests

- [X] manually tested
- [ ] added unit tests
- [ ] added integration tests
- [ ] verified on staging environment (screenshot attached)

Before:
```
% make lint
hatch run verify
cmd [1] | black --check . --extend-exclude 'tests/unit/source_code/samples/*'
All done! ✨ 🍰 ✨
276 files would be left unchanged.
cmd [2] | ruff check .
All checks passed!
cmd [3] | mypy --exclude 'tests/unit/source_code/samples/*' .
src/databricks/labs/ucx/source_code/linters/imports.py:168: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
src/databricks/labs/ucx/source_code/linters/imports.py:169: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
src/databricks/labs/ucx/source_code/linters/dbfs.py:18: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
src/databricks/labs/ucx/aws/access.py:55: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
src/databricks/labs/ucx/hive_metastore/catalog_schema.py:49: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
tests/integration/source_code/test_redash.py:6: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
tests/integration/source_code/test_redash.py:7: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
tests/unit/source_code/test_notebook.py:219: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
tests/unit/source_code/test_notebook.py:235: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
tests/integration/conftest.py:266: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
tests/integration/conftest.py:267: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
tests/integration/conftest.py:268: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
tests/integration/conftest.py:413: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
tests/integration/conftest.py:498: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
Success: no issues found in 276 source files
cmd [4] | pylint --output-format=colorized -j 0 src tests

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

After:
```
% make lint                                                                                      
hatch run verify
cmd [1] | black --check . --extend-exclude 'tests/unit/source_code/samples/*'
All done! ✨ 🍰 ✨
276 files would be left unchanged.
cmd [2] | ruff check .
All checks passed!
cmd [3] | mypy --exclude 'tests/unit/source_code/samples/*' .
Success: no issues found in 276 source files
cmd [4] | pylint --output-format=colorized -j 0 src tests

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

### CODE REVIEW

File: src/databricks/labs/ucx/hive_metastore/catalog_schema.py [hunk 1]
```diff
@@ -45,15 +45,19 @@ def _update_principal_acl(self):
...
 
-    def _get_catalog_schema_grants(self):
+    def _get_catalog_schema_grants(self) -> list[Grant]:
         catalog_grants: set[Grant] = set()
         new_grants = []
         src_trg_schema_mapping = self._get_database_source_target_mapping()
         grants = self._principal_grants.get_interactive_cluster_grants()
         # filter on grants to only get database level grants
-        database_grants = [grant for grant in grants if grant.table is None and grant.view is None]
+        database_grants = [
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/catalog_schema.py:54]: rewrite as for loop, as this has just became less readable in a list comprehension

File: src/databricks/labs/ucx/aws/access.py [hunk 2]
```diff
@@ -44,7 +44,9 @@ def __init__(
         self._aws_account_id = aws_resources.validate_connection().get("Account")
         self._kms_key = kms_key
 
-    def list_uc_roles(self, *, single_role=True, role_name="UC_ROLE", policy_name="UC_POLICY"):
+    def list_uc_roles(
+        self, *, single_role: bool = True, role_name: str = "UC_ROLE", policy_name: str = "UC_POLICY"
```
Comment (nfx) [inline src/databricks/labs/ucx/aws/access.py:48]:
```suggestion
        self, *, single_role: bool = True, role_name: str = "UC_ROLE", policy_name: str = "UC_POLICY",
```

**Review by nfx — CHANGES_REQUESTED** [review]

File: tests/unit/source_code/test_notebook.py [hunk 3]
```diff
@@ -215,7 +215,7 @@ def test_notebook_builds_python_dependency_graph(mock_path_lookup):
     assert graph.all_paths == {mock_path_lookup.cwd / path for path in paths}
 
 
-def test_detects_manual_migration_in_dbutils_notebook_run_in_python_code_():
+def test_detects_manual_migration_in_dbutils_notebook_run_in_python_code_() -> None:
```
Comment (JCZuurmond) [inline tests/unit/source_code/test_notebook.py:218]: Why add the type hint only to these tests?

Comment (asnare) [reply tests/unit/source_code/test_notebook.py:218]: In this case the situation is:

 - Mypy doesn't lint function bodies if the function definition doesn't include any type hints.
 - The other tests all have a type hint in them somewhere (normally the arguments).
 - These are no-arg tests, so unless we hint the return type they're untyped and therefore Mypy ignores them. (This is what those warnings in the PR description are all about.)

Arguably we should be fully hinting everything else as well, but it wasn't the focus of this PR which is just to ensure that most function bodies are being checked and remove the warnings during `make verify`.

Comment (asnare) [reply tests/unit/source_code/test_notebook.py:218]: … and I just noticed there are other functions in the same file without any type hints. I'm not sure why they weren't triggering the same warnings. 🤔

**Review by nfx — APPROVED** [review]

**Pull request merged**